velox/type/Filter.cpp

```diff
@@ -276,6 +285,227 @@ bool BigintValuesUsingHashTable::testInt64Range(
+  for (int i = 0; i < bitmask_.size(); i++) {
+    if (bitmask_[i]) {
+      values.push_back(min_ + i);
+    }
+  }
+  return values;
+}
+
+bool NegatedBigintValuesUsingBitmask::testInt64Range(
+    int64_t min,
+    int64_t max,
+    bool hasNull) const {
+  if (hasNull && nullAllowed_) {
+    return true;
+  }
+
+  if (min == max) {
+    return testInt64(min);
+  }
+
+  // boundary checks: if max < min_ then min < min_ as well and vice versa
+  // since part of the range is outside the bitmask, the endpoint is accepted
+  if (min < min_ || max > max_)
+    return true;
+
+  // check the range manually
+  for (int i = min - min_; i <= max - min_; ++i)
+    if (!bitmask_[i])
+      return true;
+
+  return false;
+}
+
+NegatedBigintValuesUsingHashTable::NegatedBigintValuesUsingHashTable(
+    int64_t min,
+    int64_t max,
+    const std::vector<int64_t>& values,
+    bool nullAllowed)
+    : Filter(true, nullAllowed, FilterKind::kNegatedBigintValuesUsingHashTable),
+      min_(min),
+      max_(max),
+      values_(values) {
+  constexpr int32_t kPaddingElements = 4;
+  VELOX_CHECK(min < max, "min must be less than max");
+  VELOX_CHECK(values.size() > 1, "values must contain at least 2 entries");
+
+  // Size the hash table to be 2+x the entry count, e.g. 10 entries
+  // gets 1 << log2 of 50 == 32. The filter is expected to fail often so we
+  // wish to increase the chance of hitting empty on first probe.
+  auto size = 1u << (uint32_t)std::log2(values.size() * 5);
+  hashTable_.resize(size + kPaddingElements);
+  sizeMask_ = size - 1;
+  std::fill(hashTable_.begin(), hashTable_.end(), kEmptyMarker);
+  for (auto value : values) {
+    if (value == kEmptyMarker) {
+      containsEmptyMarker_ = true;
+    } else {
+      auto position = ((value * M) & (size - 1));
+      for (auto i = position; i < position + size; i++) {
+        uint32_t index = i & sizeMask_;
+        if (hashTable_[index] == kEmptyMarker) {
+          hashTable_[index] = value;
+          break;
+        }
+      }
+    }
+  }
+  // Replicate the last element of hashTable kPaddingEntries times at 'size_' so
+  // that one can load a full vector of elements past the last used index.
+  for (auto i = 0; i < kPaddingElements; ++i) {
+    hashTable_[sizeMask_ + 1 + i] = hashTable_[sizeMask_];
+  }
+  std::sort(values_.begin(), values_.end());
+}
+
+bool NegatedBigintValuesUsingHashTable::testInt64(int64_t value) const {
+  if (containsEmptyMarker_ && value == kEmptyMarker) {
+    return false;
+  }
+  if (value < min_ || value > max_) {
+    return true;
+  }
+  uint32_t pos = (value * M) & sizeMask_;
+  for (auto i = pos; i <= pos + sizeMask_; i++) {
+    int32_t idx = i & sizeMask_;
+    int64_t l = hashTable_[idx];
+    if (l == kEmptyMarker) {
+      return true;
+    }
+    if (l == value) {
+      return false;
+    }
+  }
+  return true;
+}
+
+xsimd::batch_bool<int64_t> NegatedBigintValuesUsingHashTable::testValues(
```

Comment:
FYI, We can template this and the other testValue functions ; So we dont have to replicate for each integral type.